NativeAOT-LLVM: Compile fails due to invalid cast error

I tried to compile [Space Station 14](https://github.com/space-wizards/space-station-14/)'s client as a test run, if only just to get a ballpark of how large the final wasm binary would be. I know full-well there's no chance in hell it's going to work just like this, but while compiling the code on Release the following error did crop up:

```
EXEC : error : Invalid cast (Producer: 'LLVM17.0.4' Reader: 'LLVM 18.0.0git') [E:\.nuget\microsoft.dotnet.ilcompiler.llvm\9.0.0-alpha.1.24173.1\build\ExecWrapper.proj]
  1 error generated.
E:\.nuget\microsoft.dotnet.ilcompiler.llvm\9.0.0-alpha.1.24173.1\build\ExecWrapper.proj(3,5): error MSB3073: The command ""E:\Projects\emsdk\/upstream/bin/clang++.exe" "obj\Release\wasi-wasm\native\Content.Client.7.bc" -o "obj\Relea
se\wasi-wasm\native\Content.Client.7.o" -fvisibility=default -mllvm -combiner-global-alias-analysis=false -mllvm -disable-lsr --sysroot="D:\Downloads\wasi-sdk-20.0.m-mingw.tar\wasi-sdk-20.0.m-mingw\wasi-sdk-20.0+m\/share/wasi-sysroo
t" -target wasm32-unknown-wasi -c -O2 -g3" exited with code 1.
EXEC : error : Invalid cast (Producer: 'LLVM17.0.4' Reader: 'LLVM 18.0.0git') [E:\.nuget\microsoft.dotnet.ilcompiler.llvm\9.0.0-alpha.1.24173.1\build\ExecWrapper.proj]
  1 error generated.
E:\.nuget\microsoft.dotnet.ilcompiler.llvm\9.0.0-alpha.1.24173.1\build\ExecWrapper.proj(3,5): error MSB3073: The command ""E:\Projects\emsdk\/upstream/bin/clang++.exe" "obj\Release\wasi-wasm\native\Content.Client.4.bc" -o "obj\Relea
se\wasi-wasm\native\Content.Client.4.o" -fvisibility=default -mllvm -combiner-global-alias-analysis=false -mllvm -disable-lsr --sysroot="D:\Downloads\wasi-sdk-20.0.m-mingw.tar\wasi-sdk-20.0.m-mingw\wasi-sdk-20.0+m\/share/wasi-sysroo
t" -target wasm32-unknown-wasi -c -O2 -g3" exited with code 1.
```

Reproduction branch:
* https://github.com/PJB3005/space-station-14-content/tree/24-04-22-nativeaot-llvm
* https://github.com/PJB3005/space-station-14/tree/24-04-22-nativeaot-llvm

Note that the project makes use of submodules: the `RobusToolbox` submodule in the main repo (first link) needs to be updated to point to the second link.

Error occurs when doing `dotnet publish -r wasi-wasm -c Release /p:UseAppHost=false` inside `Content.Client` on those branches.

NativeAOT-LLVM version: 9.0.0-alpha.1.24173.1
Emscripten SDK version: 3.1.47
WASI SDK version: 22.0

--- a/src/coreclr/jit/llvmlower.cpp
+++ b/src/coreclr/jit/llvmlower.cpp
@@ -329,9 +329,18 @@ void Llvm::lowerLocal(GenTreeLclVarCommon* node)
         lowerStoreLcl(node->AsLclVarCommon());
     }
 
-    if (node->OperIsLocalStore() && node->TypeIs(TYP_STRUCT) && genActualTypeIsInt(node->gtGetOp1()))
+    if (node->TypeIs(TYP_STRUCT) && node->OperIsLocalStore())
     {
-        node->gtGetOp1()->SetContained();
+        GenTree* value = node->Data();
+        if (genActualTypeIsInt(value))
+        {
+            value->SetContained();
+        }
+        else if (node->OperIs(GT_STORE_LCL_FLD))
+        {
+            // "STORE_LCL_VAR"s already normalized by "lowerStoreLcl".
+            node->AsLclFld()->SetLayout(value->GetLayout(_compiler));
+        }
     }
 
     if (node->OperIsLocalField() || node->OperIs(GT_LCL_ADDR))

--- a/src/tests/nativeaot/SmokeTests/HelloWasm/HelloWasm.cs
+++ b/src/tests/nativeaot/SmokeTests/HelloWasm/HelloWasm.cs
@@ -415,6 +415,8 @@ internal unsafe partial class Program
 
         TestJitUseStruct();
 
+        TestMismatchedStructLocalFieldStore();
+
         TestUnsafe();
 
         TestReadByteArray();
@@ -468,6 +470,27 @@ internal unsafe partial class Program
         var res = JitUseStructProblem(&structWithStruct, structWithIndex);
 
         EndTest(res.Index == structWithIndex.Index && res.Value == structWithIndex.Value);
+    }
+
+    public struct StructWrapper
+    {
+        public readonly Guid WrappedStruct;
+        public StructWrapper(Guid value) => WrappedStruct = value;
+    }
+
+    private static void TestMismatchedStructLocalFieldStore()
+    {
+        const string GuidValue = "0c733a1e-2a1c-11ce-ade5-00aa0044773d";
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        bool ExposeAndVerify(ref StructWrapper? x)
+        {
+            return x.HasValue && x.Value.WrappedStruct.ToString() == GuidValue;
+        }
+
+        StartTest("TestMismatchedStructLocalFieldStore");
+        StructWrapper? x = new StructWrapper(Guid.Parse(GuidValue));
+        EndTest(ExposeAndVerify(ref x));
     }
 
     [StructLayout(LayoutKind.Sequential)]

